core/autocrypt/claimchain.py

```diff
@@ -0,0 +1,223 @@
+out-of-band verification of claimchain heads and key material between two users.
+
+todo/to-consider:
+
+- properly implement oob verification
+
+- look into using ipfs's modules/concepts for serializing
+  and creating "content ids", i.e. self-describing hash addresses to blocks
+
+- add crypto signing of each entry?  For in-band transmission of
+  ClaimChains we can probably just sign the whole chain instead
+  of the single entries.
+
+"""
+from __future__ import unicode_literals, print_function
+
+import os
+import time
+import json
+import hashlib
+import codecs
+from pprint import pprint
+from autocrypt import mime
+
+
+class BlockService:
+    """ Filesystem Blockservice for storing and getting immutable blocks
+    for use from ClaimChain instances. """
+    def __init__(self, basedir):
+        self._basedir = basedir
+
+    def store_block(self, type, args, parent=None):
+        # we choose the simplest data structure to create a block for a claimchain
+        # each block references a parent block (or None if it's the
+        # genesis block) and a timestamp.
+        data = [type, parent, time.time()] + list(args)
+        serialized = json.dumps(data).encode("utf8")
+        cid = hashlib.sha256(serialized).hexdigest()
+        path = os.path.join(self._basedir, cid)
+        with open(path, "wb") as f:
+            f.write(serialized)
+        return Block(cid, data, bs=self)
+
+    def get_block(self, cid, raising=True):
+        path = os.path.join(self._basedir, cid)
+        if os.path.exists(path):
+            with codecs.open(path, "r", encoding="utf-8") as f:
+                data = json.load(f)
+            return Block(cid, data, bs=self)
+        if raising:
+            raise LookupError("block not found: {r}".format(cid))
+
+
+class Block:
+    """ Basic entry for claim chains.  Each Block has the
+    following attributes:
+    - cid: the content address of this block
+    - parent_cid: the parent content address or None
+    - timestamp: when this block was created in seconds since epoch
+    - args: the block-specific payload
+    """
+    def __init__(self, cid, data, bs):
+        self.cid = cid
+        self.type = data[0]
+        self.parent_cid = data[1]
+        self.timestamp = data[2]
+        self.args = data[3:]
+        self._bs = bs
+
+    @property
+    def parent(self):
+        """ parent block or None. """
+        if self.parent_cid:
+            return self._bs.get_block(self.parent_cid)
+
+    def __eq__(self, other):
+        return self.cid == other.cid
+
+    def __iter__(self):
+        current = self
+        while current:
+            yield current
+            current = current.parent
+
+    def list_blocks(self):
+        return list(self)
+
+    def num_blocks(self):
+        return len(self.list_blocks)
+
+    def contains_cid(self, cid):
+        for x in self:
+            if x.cid == cid:
+                return True
+
+    def get_first_block(self):
```

Comment:
minor: This confused me because i thought this would be the same as the first block i get when iterating through them - which would be self if i understand correctly.
get_genisis_block / get_oldest_block might be more clear.